### PR TITLE
Avoid fetching suggestions all the time

### LINKED ISSUE #3030: Autocomplete files when clicking on an already-run search
### Describe the problem and steps to reproduce it:
If you type a search term into the search box, let the results load, click away, and then focus the search again, it will run the autocomplete search a second time despite having local results. This creates an unneeded network request and also feels a bit too frantic.

### What happened?
The autocomplete action dispatched again despite the filters all being the same.

### What did you expect to happen?
The existing results are simply loaded/shown again because the filters haven't changed.

### PR DESCRIPTION
**Pull request by willdurand**
Fix #3030 

---

This PR slightly changes the autocomplete feature to avoid too many API calls.

This was more complicated than I thought 😞 Hopefully the new test cases are meaningful enough but they should cover all the edge cases I encountered.

### CODE REVIEW

File: src/amo/components/SearchForm/index.js [hunk 1]
```diff
@@ -55,6 +57,7 @@ export class SearchFormBase extends React.Component {
 
     this.state = {
       searchValue: props.query || '',
+      lastAutocompleteFilters: {},
```
Comment (tofumatt) [inline src/amo/components/SearchForm/index.js]: Could this be either `previousAutoCompleteFilters` or `oldFilters`? Last sounds like the latest and is confusing, so I think this could be better named.

File: src/amo/components/SearchForm/index.js [hunk 2]
```diff
@@ -119,10 +123,22 @@ export class SearchFormBase extends React.Component {
       filters.addonType = addonType;
     }
 
-    dispatch(autocompleteStart({
-      errorHandlerId: errorHandler.id,
-      filters,
-    }));
+    if (
+      // See: https://github.com/moroshko/react-autosuggest#onsuggestionsfetchrequested-required.
```
Comment (tofumatt) [inline src/amo/components/SearchForm/index.js]: Thanks for linking to that, very handy 👍

File: src/amo/components/SearchForm/index.js [hunk 3]
```diff
@@ -119,10 +123,22 @@ export class SearchFormBase extends React.Component {
...
+      reason === 'input-focused' &&
+      deepEqual(filters, this.state.lastAutocompleteFilters)
+    ) {
+      log.debug(oneLine`Ignoring suggestions fetch requested because it has been
```
Comment (tofumatt) [inline src/amo/components/SearchForm/index.js]: I think this could be a bit clearer/concise, maybe just:

"No autocomplete start dispatched: search input focused but filters have not changed."

File: tests/unit/amo/components/TestSearchForm.js [hunk 4]
```diff
@@ -219,6 +219,173 @@ describe(__filename, () => {
...
+        dispatch,
+        suggestions: autocompleteState.suggestions,
+      });
+
+      sinon.assert.notCalled(dispatch);
+
+      // User types 'a' in the search input.
+      wrapper.find('input').simulate('change', createFakeChangeEvent('a'));
+
+      // User clicks somewhere else on the UI, triggering the
```
Comment (tofumatt) [inline tests/unit/amo/components/TestSearchForm.js]: These comments in the test describing the actions the user would take are fantastic, thank you much!

File: tests/unit/amo/components/TestSearchForm.js [hunk 5]
```diff
@@ -219,6 +219,173 @@ describe(__filename, () => {
...
+
+      // User focuses the search input.
+      wrapper.find('input').simulate('focus');
+
+      expect(wrapper.find(Suggestion)).toHaveLength(
+        autocompleteState.suggestions.length
+      );
+
+      sinon.assert.callCount(dispatch, 2);
+      sinon.assert.calledWith(dispatch, autocompleteStart({
```
Comment (tofumatt) [inline tests/unit/amo/components/TestSearchForm.js]: I think this test would be ever more robust if you reset the dispatch before `wrapper.setProps({ addonType: ADDON_TYPE_THEME });`. Technically this isn't checking the call order, which I know is minor but if you instead made sure it's called once after each event we expect the dispatch on we can check the order too.

It's super minor but I think it's better overall.

File: tests/unit/amo/components/TestSearchForm.js [hunk 6]
```diff
@@ -219,6 +219,173 @@ describe(__filename, () => {
...
+
+      // User types 'adb' in the search input, which requires new suggestions
+      // to be loaded from the API.
+      wrapper.find('input').simulate('change', createFakeChangeEvent('adb'));
+
+      expect(wrapper.find(Suggestion)).toHaveLength(
+        autocompleteState.suggestions.length
+      );
+
+      sinon.assert.callCount(dispatch, 2);
```
Comment (tofumatt) [inline tests/unit/amo/components/TestSearchForm.js]: Same here; I think doing the check before `wrapper.find('input').simulate('change', createFakeChangeEvent('adb'));` and then doing `dispatch.reset()` then checking again would be a bit more robust and also more clearly show which assertion is related to specific actions we're simulating in the test.

File: src/amo/components/SearchForm/index.js (hunk 1 above)
Comment (willdurand) [reply src/amo/components/SearchForm/index.js]: Would `oldAutocompleteFilters` work for you? I started with `lastDispatched(Autocomplete)Filters`, but that was waaaay too long, so I shortened the name, but agreed that now it feels weird to have `last`.

Comment (tofumatt) [reply src/amo/components/SearchForm/index.js]: `oldAutocompleteFilters` works for me; we use `oldFilters` elsewhere. I think `previous` is better still but you're right it's verbose and we're already using the `old` prefix elsewhere so let's do that.

File: src/amo/components/SearchForm/index.js [hunk 7]
```diff
@@ -126,10 +128,22 @@ export class SearchFormBase extends React.Component {
...
+      // See: https://github.com/moroshko/react-autosuggest#onsuggestionsfetchrequested-required.
+      reason === 'input-focused' &&
+      deepEqual(filters, this.state.oldAutocompleteFilters)
+    ) {
+      log.debug(oneLine`No autocomplete start dispatched: search input focused
+        but filters have not changed.`);
+      return;
+    }
+
+    this.setState({ oldAutocompleteFilters: filters }, () => {
```
[the remark this answers is not in the document]
Comment (willdurand) [reply src/amo/components/SearchForm/index.js]: Ok I am going to remove the async callback.
